### PR TITLE
New Strategies -- CycleHunter and multiple Meta strategies

### PR DESCRIPTION
**Pull request by marcharper**
Also: improved the Meta player tests and classifier handling, and added some additional tests.

I will rebase post #340...

### CODE REVIEW

File: axelrod/strategies/hunter.py [hunk 1]
```diff
@@ -58,6 +60,38 @@ def strategy(self, opponent):
         return 'C'
 
 
+class CycleHunter(Player):
```
Comment (drvinceknight) [inline axelrod/strategies/hunter.py]: Could you throw in some docstrings for the new strategies?

**Comment by marcharper** [comment]
Note that the new Meta strategies increase the tournament running time (which is why Travis fails)

**Comment by drvinceknight** [comment]
> Note that the new Meta strategies increase the tournament running time (which is why Travis fails)

As @meatballs has just suggested on gitter, perhaps in this PR change the .travis.yml file to only run the unit tests? It would be good to include something in the contributing docs about 'please' run the functional tests locally also or something...

**Comment by marcharper** [comment]
We could also just drop the meta strategies from the full tournament test or otherwise shorten it (maybe less than 500 rounds would help).

**Comment by drvinceknight** [comment]
We could go for something trivially short? 20 rounds?

**Comment by marcharper** [comment]
For some reason it's the "all strategies" tournament that takes dramatically longer. Here's a run on the tournament repository:
```
-- Regenerating noisy tournament figures
Starting basic_strategies tournament with 10 round robins of 200 turns per pair.
Finished all basic_strategies tasks in 1.6s
Starting strategies tournament with 10 round robins of 200 turns per pair.
Finished all strategies tasks in 137.9s
Starting cheating_strategies tournament with 10 round robins of 200 turns per pair.
Finished all cheating_strategies tasks in 8.9s
Starting all_strategies tournament with 10 round robins of 200 turns per pair.
Finished all all_strategies tasks in 1679.4s
Finished all tournaments in 1827.7s
-- Running Axelrod, regenerating cache, results, and figures
Starting basic_strategies tournament with 10 round robins of 200 turns per pair.
Finished all basic_strategies tasks in 1.5s
Starting strategies tournament with 10 round robins of 200 turns per pair.
Finished all strategies tasks in 115.0s
Starting cheating_strategies tournament with 10 round robins of 200 turns per pair.
Finished all cheating_strategies tasks in 3.9s
Starting all_strategies tournament with 10 round robins of 200 turns per pair.
Finished all all_strategies tasks in 2783.6s
Finished all tournaments in 2904.0s
Be sure to git commit changes to update the docs!
```

**Comment by drvinceknight** [comment]
> The noise-free tournament is actually faster -- shouldn't that not be the case?

That is strange.

What about the option of just reducing the rounds length (drastically). It's also a good test of non default round lengths...

**Comment by marcharper** [comment]
> What about the option of just reducing the rounds length (drastically). It's also a good test of non default round lengths...

Yeah I like that option -- I don't see a big benefit to running for more rounds in the test. 

But we should investigate further...

**Comment by drvinceknight** [comment]
> Yeah I like that option -- I don't see a big benefit to running for more rounds in the test.

:+1: 

> 
> But we should investigate further...

:+1:

**Comment by marcharper** [comment]
[Profiler output for run_axelrod](https://docs.google.com/spreadsheets/d/1hcNnz5XTmn4AuE9NfRxiUkOe-47BNioM923Pa0GM6Qk/edit?usp=sharing)

As you can see, the vast majority of the time spent appears to be in the `recv` method of multiprocessing. I'm not sure if this method is simply hiding all the actual work or if there is more to it, however I suspect there is some issue related to the underlying OS pipe size or some other problem with multiprocessing communication. Note that the time really increases dramatically for the "all_strategies" tournament -- the smaller tournaments do not seem to invoke this problem.

[This thread](https://github.com/datamade/dedupe/issues/305) may be relevant.

Some fun facts from the profiler:
* The tournaments require 24 million random numbers
* 12 million applications of math.log
* 587k `isinstance` checks

**Comment by drvinceknight** [comment]
I was just listening to a podcast about profiling: neat!

Not sure I can add anything helpful, this is very much at the edge of my understanding of parallel processing...

**Comment by marcharper** [comment]
I should note that I ran a noisy tournament, which undoubtedly inflated the random number count:
```
python -m cProfile run_axelrod --noise 0.05 --rc -p 0 -o assets > profile.txt
```

**Comment by marcharper** [comment]
Well it's not due to the number of players. A tournament with 500 Random players runs quickly.

**Comment by marcharper** [comment]
I think you'll find the [single thread output](https://docs.google.com/spreadsheets/d/1N_TVIwD9tO2u3JpN2BSFwuYFAtYXjh3Tttd2eWbVTHM/edit?usp=sharing) informative. It took 1500 fewer seconds!

MindReader's `lookahead` function takes 1600 seconds (not surprising since it uses `deepcopy`) of the ~1800 seconds that the full tournament uses. That explains most of the difference, and it just takes much longer with more strategies (and coupled with the hunter and meta strategies, which are on the higher end of time consuming, and can take a long time to copy).

```
$ python -m cProfile run_axelrod --noise 0.05 --rc -o assets > single_profile.txt

Starting basic_strategies tournament with 10 round robins of 200 turns per pair.
Finished all basic_strategies tasks in 2.2s
Starting strategies tournament with 10 round robins of 200 turns per pair.
Finished all strategies tasks in 563.2s
Starting cheating_strategies tournament with 10 round robins of 200 turns per pair.
Finished all cheating_strategies tasks in 27.2s
Starting all_strategies tournament with 10 round robins of 200 turns per pair.
Finished all all_strategies tasks in 1836.0s
Finished all tournaments in 2428.6s
```

**Comment by marcharper** [comment]
Rebased onto upstream and reduced the integration test lengths.

File: axelrod/tests/unit/test_meta.py [hunk 2]
```diff
@@ -3,13 +3,51 @@
...
+import unittest
 
 from .test_player import TestPlayer
 
 C, D = 'C', 'D'
 
 
-class TestMetaMajority(TestPlayer):
+class TestMetaPlayer(TestPlayer):
+
```
Comment (drvinceknight) [inline axelrod/tests/unit/test_meta.py]: Could you throw in a docstring here?

File: axelrod/strategies/calculator.py [hunk 3]
```diff
@@ -23,20 +23,20 @@ def __init__(self):
...
-    @staticmethod
-    def detect_cycle(history):
-        """Detects if there is a cycle in the opponent's history."""
-        for i in range(len(history) // 2):
-            cycle = itertools.cycle(history[0: i + 1])
-            cycle_list = list(itertools.islice(cycle, 0, len(history)))
-            if list(history) == cycle_list:
-                return True
-        return False
+    #@staticmethod
```
Comment (drvinceknight) [inline axelrod/strategies/calculator.py]: Is this something you meant to get rid off?

Comment (marcharper) [inline axelrod/strategies/calculator.py]: indeed

File: axelrod/tests/unit/test_meta.py (hunk 2 above)
Comment (marcharper) [inline axelrod/tests/unit/test_meta.py]: Done

**Comment by drvinceknight** [comment]
Assuming Travis: :+1:

**Pull request merged**